internal/agent/runtime/docker.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"github.com/go-logr/logr"
+	"github.com/tinkerbell/tink/internal/agent"
+	"github.com/tinkerbell/tink/internal/agent/runtime/internal"
+	"github.com/tinkerbell/tink/internal/agent/workflow"
+	"github.com/tinkerbell/tink/internal/ptr"
+)
+
+var _ agent.ContainerRuntime = &Docker{}
+
+// Docker is a docker runtime that satisfies agent.ContainerRuntime.
+type Docker struct {
+	log    logr.Logger
+	client *client.Client
+}
+
+// Run satisfies agent.ContainerRuntime.
+func (d *Docker) Run(ctx context.Context, a workflow.Action) error {
+	pullImage := func() error {
+		// We need the image to be available before we can create a container.
+		image, err := d.client.ImagePull(ctx, a.Image, types.ImagePullOptions{})
+		if err != nil {
+			return fmt.Errorf("docker: %w", err)
+		}
+		defer image.Close()
+
+		// Docker requires everything to be read from the images ReadCloser for the image to actually
+		// be pulled. We may want to log image pulls in a circular buffer somewhere for debugability.
+		if _, err = io.Copy(io.Discard, image); err != nil {
+			return fmt.Errorf("docker: %w", err)
+		}
+
+		return nil
+	}
+
+	err := retry.Do(pullImage, retry.Attempts(5), retry.DelayType(retry.BackOffDelay))
+	if err != nil {
+		return err
+	}
+
+	// TODO: Support all the other things on the action such as volumes.
+	cfg := container.Config{
+		Image: a.Image,
+		Env:   toDockerEnv(a.Env),
+	}
+
+	failureFiles, err := internal.NewFailureFiles()
+	if err != nil {
+		return fmt.Errorf("create action failure files: %w", err)
+	}
+	defer failureFiles.Close()
+
+	hostCfg := container.HostConfig{
+		Mounts: []mount.Mount{
+			{
+				Type:   mount.TypeBind,
+				Source: failureFiles.ReasonPath(),
+				Target: ReasonMountPath,
+			},
+			{
+				Type:   mount.TypeBind,
+				Source: failureFiles.MessagePath(),
+				Target: MessageMountPath,
+			},
+		},
+	}
+
+	containerName := toContainerName(a.ID)
+
+	// Docker uses the entrypoint as the default command. The Tink Action Cmd property is modeled
+	// as being the command launched in the container hence it is used as the entrypoint. Args
+	// on the action are therefore the command portion in Docker.
+	if a.Cmd != "" {
+		cfg.Entrypoint = append(cfg.Entrypoint, a.Cmd)
+	}
+	if len(a.Args) > 0 {
+		cfg.Cmd = append(cfg.Cmd, a.Args...)
+	}
+
+	// TODO: Figure out container logging. We probably want to save it somewhere for debugability.
+
+	create, err := d.client.ContainerCreate(ctx, &cfg, &hostCfg, nil, nil, containerName)
+	if err != nil {
+		return fmt.Errorf("docker: %w", err)
+	}
+
+	// Always try to remove the container on exit.
+	defer func() {
+		// Force remove containers in an attempt to preserve space in memory constraints environments.
+		// In rare cases this may create orphaned volumes that the Docker CLI won't clean up.
+		opts := types.ContainerRemoveOptions{
+			Force:       true,
+			RemoveLinks: true,
```

Comment:
This is causing issues in tests and its not clear why. I no longer think I understand what it does so I'm disabling; we can enable it at a later date.